File: backend/src/main/java/com/example/backend/service/UserService.java
package com.example.backend.service;

import com.example.backend.model.User;
import com.example.backend.repository.UserRepository;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

@Service
public class UserService {

    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;

    public UserService(UserRepository userRepository, PasswordEncoder passwordEncoder) {
        this.userRepository = userRepository;
        this.passwordEncoder = passwordEncoder;
    }

    public User getById(String userId) {
        return userRepository.findById(userId)
                .orElseThrow(() -> new IllegalArgumentException("User not found."));
    }

    public User updateProfile(String userId, String name, String avatar, String bio) {
        User user = getById(userId);
        if (name != null && !name.isBlank()) {
            user.setName(name);
        }
        if (avatar != null) {
            user.setAvatar(avatar);
        }
        if (bio != null) {
            user.setBio(bio);
        }
        return userRepository.save(user);
    }

    public void changePassword(String userId, String currentPassword, String newPassword) {
        User user = getById(userId);
        if (!passwordEncoder.matches(currentPassword, user.getPassword())) {
            throw new IllegalArgumentException("Current password is incorrect.");
        }
        user.setPassword(passwordEncoder.encode(newPassword));
        userRepository.save(user);
    }
}

File: backend/src/services/userService.js
import bcrypt from "bcryptjs";
import { User } from "../models/user.js";

export async function getById(userId) {
  const user = await User.findById(userId);
  if (!user) {
    throw new Error("User not found.");
  }
  return user;
}

export async function updateProfile(userId, name, avatar, bio) {
  const user = await getById(userId);
  if (name?.trim()) {
    user.name = name;
  }
  if (avatar !== undefined) {
    user.avatar = avatar;
  }
  if (bio !== undefined) {
    user.bio = bio;
  }
  return user.save();
}

export async function changePassword(userId, currentPassword, newPassword) {
  const user = await getById(userId);
  if (!(await bcrypt.compare(currentPassword || "", user.password || ""))) {
    throw new Error("Current password is incorrect.");
  }
  user.password = await bcrypt.hash(newPassword, 10);
  await user.save();
}

export function safeUserPayload(user) {
  return {
    id: user.id,
    name: user.name || "",
    email: user.email || "",
    avatar: user.avatar || "",
    bio: user.bio || "",
  };
}
